<!--BOOK_INFORMATION-->
*This notebook contains an excerpt from the [Python Programming and Numerical Methods - A Guide for Engineers and Scientists](https://www.elsevier.com/books/python-programming-and-numerical-methods/kong/978-0-12-819549-9), the content is also available at [Berkeley Python Numerical Methods](https://pythonnumericalmethods.berkeley.edu/notebooks/Index.html).*

*The copyright of the book belongs to Elsevier. We also have this interactive book online for a better learning experience. The code is released under the [MIT license](https://opensource.org/licenses/MIT). If you find this content useful, please consider supporting the work on [Elsevier](https://www.elsevier.com/books/python-programming-and-numerical-methods/kong/978-0-12-819549-9) or [Amazon](https://www.amazon.com/Python-Programming-Numerical-Methods-Scientists/dp/0128195495/ref=sr_1_1?dchild=1&keywords=Python+Programming+and+Numerical+Methods+-+A+Guide+for+Engineers+and+Scientists&qid=1604761352&sr=8-1)!*

# Reading
Python Programming and Numerical Methods, 
[Chapter 9. Representation of Numbers](https://pythonnumericalmethods.studentorg.berkeley.edu/notebooks/chapter09.00-Representation-of-Numbers.html)

# Problems

## 1. Write a function *my_bin_2_dec(b)*
where *b* is binary number represented by a list of ones and zeros. The last element of *b* represents the coefficient of $2^0$, the second-to-last element of b represents the coefficient of $2^1$, and so on. The output variable, *d*, should be the decimal representation of b. The test cases are provided below. 

In [88]:
def my_bin_2_dec(bin):
    dec = 0
    
    arr = len(bin)
    for i in range(arr):
        dec = dec + ((bin[-arr + i]) * (2**(arr-1-i)))      # The negative index allows the loop to travverse the array in the other direction
    return dec

In [89]:
# Output: 7
my_bin_2_dec([1, 1, 1])

7

In [90]:
# Output: 85
my_bin_2_dec([1, 0, 1, 0, 1, 0, 1])

85

In [91]:
# Output: 33554431
my_bin_2_dec([1]*25)

33554431

## 2. Write a function *my_dec_2_bin(d)*
where *d* is a positive integer in decimal, and *b* is the binary representation of *d*. The output *b* must be a list of ones and zeros, and the leading term must be a 1 unless the decimal input value is 0. The test cases are provided below. 

In [92]:
import numpy as np

def my_dec_2_bin(decimal):
    if(decimal == 0):
        return np.array([0])

    bits = int(np.floor(np.log2(decimal)) + 1)      # Set how many bits are needed
    binary = np.zeros(bits)
    
    for i in range(bits - 1, -1, -1):               # This loop goes backwards -> l-1 to -1
        binary[i] = decimal % 2                     # Takes the remainder and if it is 1, make that the bit value. If it is 0, make that the bit value
        decimal = decimal // 2                      # Removes that bit's contribution (divides by two) and then rounds down (we took care of the remainder)

    return binary

In [93]:
# Output: [0]
my_dec_2_bin(0)

array([0])

In [94]:
# Output: [1, 0, 1, 1, 1]
my_dec_2_bin(23)

array([1., 0., 1., 1., 1.])

In [95]:
# Output: [1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1]
my_dec_2_bin(2097)

array([1., 0., 0., 0., 0., 0., 1., 1., 0., 0., 0., 1.])

## 3. Compute *d = my_bin_2_dec(my_dec_2_bin(12654))*
Use the two functions you wrote in problems 1 and 2. Do you get the same number?

In [96]:
d = my_bin_2_dec(my_dec_2_bin(12654))
print(d)

12654.0


## 4. Write a function *my_bin_adder(b1,b2)*
where *b1*, *b2* and the output variable *b* are binary numbers represented as in problem 1. The output variable should be computed as *b = b1 + b2*. Do not use your functions from problems 1 and 2 to write this function (i.e., do not convert *b1* and *b2* to decimal; add them, and then convert the result back to binary). This function should be able to accept inputs *b1* and *b2* of any length (i.e., very long binary numbers), and *b1* and *b2* may not necessarily be the same length.

In [97]:
def my_bin_adder(b1, b2):
    ## Since we start adding with the LSB, we need to index our arrays from right to left
    index_1 = len(b1) - 1
    index_2 = len(b2) - 1
    answer = []
    carry_bit = 0

    while ((index_1 >= 0) or (index_2 >= 0) or carry_bit):      # This is a loop that runs as long as there are numbers to add basically. Technically, it checks if theses values are != 0
        # Here, I will store the bit in a variable if the cooresponding binary number still has bits left to add
        if index_1 >= 0:
            bit_1 = b1[index_1]
        else:
            bit_1 = 0
        if index_2 >= 0:
            bit_2= b2[index_2]
        else:
            bit_2 = 0

        # Add the three bits together. Possible results: 0, 1, or 2. If 0, add a 0 into the answer. If 1, add the 1 in the answer. If 2, add 0 in the answer but store a 1 in the carry. 
        added = bit_1 + bit_2 + carry_bit
        answer.append(added % 2)                # This will add a 0 if added was 0 or 2, and a 1 if added was 1
        carry_bit = added // 2                  # This will store a 1 in carry if added was 2, but store a 0 if added was 0 or 1

        # Decrement our index values towards the MSB
        index_1 = index_1 - 1
        index_2 = index_2 - 1

    ## Weird bug... The array "answer" was built backwards. Luckily, in all my wisdom, instead of manually flipping the array around, I found a function to do it for me
    answer.reverse()
    return answer


In [98]:
# Output: [1, 0, 0, 0, 0, 0]                    ## I think that this output is wrong in the comment
my_bin_adder([1, 0], [1, 1, 1, 1, 1])

[1, 0, 0, 0, 0, 1]

In [99]:
# Output: [1, 1, 1, 0, 0, 1, 1]
my_bin_adder([1, 1, 1, 1, 1], [1, 0, 1, 0, 1, 0, 0])

[1, 1, 1, 0, 0, 1, 1]

In [100]:
# Output: [1, 0, 1, 1]
my_bin_adder([1, 1, 0], [1, 0, 1])

[1, 0, 1, 1]

## 5. What is the effect of allocating more bits to the fraction versus the characteristic and vice versa? What is the effect of allocating more bits to the sign?

### Characteristic  --> Bits --> Fraction
Moving bits from the characteristic will decrease the range of the exponent and therefore decrease the range of numbers that could be reached using IEEE754. 
Adding bits to the fraction would allow a more precise approximation of a value.

### Fraction --> Bits --> Characteristic
Opposite case from above: Moving bits from the fraction decreases the precision because of the limited number of negatives powers to raise 2 to. 
Adding bits to the characteristic and taking them from the fraction would increase the range of numbers that can be represented because the exponent value could increase.

### Adding bits to the sign
This is just **not** optimal. The sign of a number is binary. It should only have 1 bit. 

## 6. Write a function *my_ieee_2_dec(ieee)*
where *ieee* is a string contains 64 characters of ones and zeros representing a 64-bit IEEE754 number. The output should be *d*, the equivalent decimal representation of *ieee*. The input variable *ieee* will always be a 64-element string of ones and zeros defining a 64-bit float. 

In [101]:
def my_ieee_2_dec(ieee):
    # SIGN
    if ieee[0] == '1': 
        sign = -1
    else:
        sign = 1
    
    # EXPONENT
    exponent = int(ieee[1:12], 2)      
        # Looked up this function to convert things to integer: int(value, base). And how to access the elements arr[start:stop(doesnt include)]

    # FRACTION
    frac_bits = ieee[12:]                       # Elements 12 to the end
    fraction = 0.0
    for i, frac_bit_1 in enumerate(frac_bits):  # Do a loop for every element in the fraction bits
        if frac_bit_1 == '1':                   # When there is a 1...
            fraction = fraction + (2**-(i+1))   # Do 2^-(i+1) and add it

    # Special Cases - Subnormal 
    # 'when e = 0 (i.e., e = 00000000000 (base2)) and when e = 2047 (i.e., e = 11111111111 (base2))'
    if exponent == 0:                           # Subnormal number, which is computed by n=(−1)^s*2^−1022*(0+f)
        return sign * fraction * (2 ** -1022)

    # When the exponent is 2047, then f = 0 and s = 0, result is positive infinity. When the exponent is 2047, then f = 0, and s = 1, result is minus infinity.
    if exponent == 2047:
        if fraction == 0:
            return sign * float('inf')
        # When the exponent is 2047 and f is nonzero, then the result is “Not a Number”, which means that the number is undefined. 
        else:
            return float('undefined')

    # Normal
    return sign * (1+fraction) * (2 ** (exponent-1023))

In [102]:
# Output: -48
ieee = '1100000001001000000000000000000000000000000000000000000000000000'
my_ieee_2_dec(ieee)

-48.0

In [103]:
# Output: 3.39999999999999991118215802999
ieee = '0100000000001011001100110011001100110011001100110011001100110011'
print(format(my_ieee_2_dec(ieee), '.30g'))          # ChatGPT had to tell me how to get it to print more decimals

3.39999999999999991118215802999


## 7. Write a function *my_dec_2_ieee(d)*
where *d* is a number in decimal and output variable *ieee* is a string with 64 characters of ones and zeros representing the 64-bit IEEE754 closest to *d*. You can assume that *d* will not cause an overflow for 64-bit *ieee* numbers.

In [104]:
import math
def my_dec_2_ieee(d):
    ## d = 0
    if d == 0.0:
        return '0' + '0'*11 + '0'*52
    
    ## SIGN  -> will already be a string
    if d < 0:
        sign = '1'
    else:
        sign = '0'
    d = abs(d)

    ## EXPONENT
    # So I did a little google and theres a really nice math function the python has. 
    # It will return the number in base 2 exponential form. [m, e] where d = m*2^e
    m, e = math.frexp(d)                # d = m * 2^e, where m is between 0.5 and 1 -> that will need to change
    exponent = e - 1 + 1023             # Because 1023 is subtracted from the exponent to normalize it, and the leading one is dropped
    
    ## Weird Cases
    if exponent <= 0:                   # Subnormal number: (-1)^s * 2^-1022 * 0.f
        # We know the sign and the exponent automatically. Now we just calculate the fraction
        fraction = int(round( (d / (2**-1022)) * (2**52) ))     # 0.f = (d/2^-1022)     f = 0.f * (2^52)
        return sign + '00000000000' + f"{fraction:052b}"
    

    ## FRACTION
    frac = m*2                          # I need to change m from 0.5<m<1 to 1<m<2. so m2^e = (2m)2^(e-1)
    frac = frac - 1                     # Frac was a number between 1 and 2, but we remove the leading 1 (because it is always there)
    fraction = int(round(frac * 2**52)) # This shifts the digets over to the left of the decimal
    # Overflow of the rounding
    if fraction == (2**52):             # If this is the case, we need to increase the exponent by 1, and 0 the fraction
        fraction = 0
        exponent = exponent + 1

    exp_str = f"{exponent:011b}"        # https://www.w3schools.com/python/python_string_formatting.asp
    frac_str = f"{fraction:052b}"       # https://www.w3schools.com/python/python_string_formatting.asp

    return sign + exp_str + frac_str

In [105]:
# Output: '0100000000101110010111101010001110011100001100011010010001101000'

d = 1.518484199625
my_dec_2_ieee(d)

'0011111111111000010010111011011000010110100011100001110100100000'

I think the given Output is wrong, I checked online too. 


In [106]:
# Output: '1100000001110011010100100100010010010001001010011000100010010000'

d = -309.141740
my_dec_2_ieee(d)

'1100000001110011010100100100010010010001001010011000100010010000'

In [107]:
# Output: '1100000011011000101010010000000000000000000000000000000000000000'

d = -25252
my_dec_2_ieee(d)

'1100000011011000101010010000000000000000000000000000000000000000'

## 8. Define *ieee_baby*
to be a representation of numbers using 6 bits where the first bit is the sign bit, the second and third bits are allocated to the characteristic, and the fourth, fifth, and sixth bits are allocated to the fraction. The normalization for the characteristic is 1.

 - Write all the decimal numbers that can be represented by *ieee_baby*. 
 - What is the largest/smallest gap in *ieee_baby*?

In [108]:
def ieee_baby(decimal):
    ## 0 case
    if decimal == 0:
        return '000000'
    

    ## SIGN
    if decimal < 0:
        sign_bits = '1'
    else:
        sign_bits = '0'
    decimal = abs(decimal)


    ## EXPONENT
    m, exponent = math.frexp(decimal)
    #exponent = exponent - 1             # Subtract one because later we multiply the m by 2
    #exponent = exponent + 1             # Add one because we always bias the exponent so we can handle different numbers


    ## FRACTION
    frac = m * 2                          # I need to change m from 0.5<m<1 to 1<m<2. so m2^e = (2m)2^(e-1)
    frac = frac - 1                     # Frac was a number between 1 and 2, but we remove the leading 1 (because it is always there)
    fraction = int(round(frac * 2**3)) # This shifts the digets over to the left of the decimal
    # Overflow of the rounding
    if fraction == (2**3):             # If this is the case, we need to increase the exponent by 1, and 0 the fraction
        fraction = 0
        exponent = exponent + 1

    characteristic_bits = f"{exponent:02b}"
    fraction_bits = f"{fraction:03b}"

    return sign_bits + characteristic_bits + fraction_bits


In [126]:
## Testing the ieee_baby a little bit
print(ieee_baby(1.25))
print(ieee_baby(-4.5))
print(ieee_baby(-0.5))
print(ieee_baby(-1.5))
print(ieee_baby(2.25))
print(ieee_baby(7.5))

001010
111001
100000
101100
010001
011111


### The system for ieee_baby is similar to the real ieee
- decimal = (-1)^s * 2^(cc-1) * (1 + f2^-1 + f2^-2 + f2^-3)
- where the ieee_baby format is sccfff. 
  - s is the sign bit
  - cc is the characteristic
  - fff is the fraction bits
  
### Numbers represented:
- ±0.5, ±0.625, ±0.75, ±0.875, ±1.0, ±1.125, ±1.25, ±1.375, ±1.5, ±1.625, ±1.75, ±1.875,
- ±2.0, ±2.25, ±2.5, ±2.75, ±3.0, ±3.25, ±3.5, ±3.75
- ±4.0, ±4.5, ±5.0, ±5.5, ±6.0, ±6.5, ±7.0, ±7.5

### Largerst/Smallest Gap
- Smallest gap: 0.125
- Largest gap: 0.5

## 9. Use the *np.spacing* function to determine the smallest number such that the gap is 1.

In [136]:
## I assume that this is for the regular float-64bit numbers? I am not sure how else it would work. 
np.spacing(1)
## Since the spacing here at 1 is super small, I will go ahead and start the search agressively, then refine
x = 1
while np.spacing(x) < 1:
    x = x * 2

## Implement binary search algorithm // Creds to CSC 1310, April Crockett
high = x                            # Spacing is >= 1
low  = x // 2                       # Spacing is < 1

while (high - low) > 1:             # Do this until the numbers are beside each other
    middle = (high+low) / 2         # midpoint
    if (np.spacing(middle) < 1):    # If the middle is still too low...
        low = middle                # Assign the midpoint to low
    else:                           # Otherwise...
        high = middle               # Assign the midpoint to high because the spacing was too big. 


print("The smallest number such that the gap is one was:", high)
print("The spacing at that point was", np.spacing(high))

The smallest number such that the gap is one was: 4503599627370496
The spacing at that point was 1.0


## 10. What are some of the advantages and disadvantages of using binary versus decimal?

#### Advantages
- Hardware freindly. 
    - There are very easy, efficent ways to use voltage and electric circuits to represent on and off, high and low, 0 and 1
    - Avoids error. With only two possible states, the reliability of reading the correct signal is high
    - Works well for logic operations
    - Powers of 2 can be pretty accurate for a lot of numbers

#### Disadvantages
- Expanded notation. Most numbers take more digits to represent than when using more characters (e.g. 0-9 or 0 -F)
- Not readable. Tons of 0's and 1's are not easy to read and understand for humans
- Not exact. Although there are tricks like ieee, the limited form of binary numbers hinders the accuracy. 

## 11. Write the number 13 (base10) in base1. How would you do addition and multiplication in base1?

0001 1111 1111 1111 (base 1) = 13 (base 10)

Addition works the same as addition worked when cavemen did it with sticks. Add however many digits to however else many digits you want to add:
- 2 + 3 (base 10) = 0011 +0111 (base 1) = 0001 1111 (base 1)
- 5 + 7 (base 10) = 11111 + 1111111 = 1111111111111 (base 1) = same as before 13 (base 10)

Multiplication works as repeated addition. For each digit of one number, you take that many copies of the other number:
- 2 * 3 (base 10)  =  11 * 111 (base 1)  =  111 + 111  =  111111 (base 1)  =  6 (base 10)
- 6 * 4 (base 10)  =  111111 * 1111  =  1111 + 1111 + 1111 + 1111 + 1111 + 1111  =  1111 1111 1111 1111 1111 1111 (base 1)

## 12. How high can you count on your fingers if you count in binary?

11111 11111(base 2) = 2^9 + 2^8 + 2^7.... + 2^1 + 2^0 = 2^10 - 1 = 1023(base 10)

## 13. Let *b* be a binary number having *n* digits. Can you think of ways to multiply and divide *b* by 2 that does not involve any arithmetic? Hint: Think about how you multiply and divide a decimal number by 10.

To multiply a binary number by a multiple of 2, we use a logic shift left (LSL).
To divide a binary number by a multiple of 2, we use a logic shift right (LSR). 

When shifted n digits left (and filled in on the right by zeros), the number is multiplied by 2^n
When shifted n digits right (and the right-most digit is erased), the number is divided by 2^n

Take the number 11 (base 10) for example:
11 * 8 (base 10) = 1011 << 3 (base 2, LSL) = 1011000 (base 2) = 88 (base 10)
11 / 2 (base 10) = 1011 >> 1 (base 2, LSR) = 101 (base 2) = 5 (base 10)     Note: the decimals are discarded in binary division unless special care is taken